oak_functions/loader/src/server.rs

```diff
@@ -740,21 +793,113 @@ mod tests {
-            .get(&ChannelHandle::LookupData)
+            .get_mut(&ChannelHandle::LookupData)
             .is_some());
     }
+
+    #[tokio::test]
+    async fn test_hosted_channel_read_from_handle_out_of_range() {
+        let mut wasm_state = create_test_wasm_state();
+        let result = wasm_state.channel_read(-1, 0, 0);
+        assert!(result.is_err());
+        assert_eq!(ChannelStatus::ChannelHandleInvalid, result.unwrap_err())
+    }
+
+    #[tokio::test]
+    async fn test_hosted_channel_read_from_handle_without_endpoint() {
+        let mut wasm_state = create_test_wasm_state();
+        // Assumes ChannelHandle 0 will never have an Endpoint.
+        let result = wasm_state.channel_read(0, 0, 0);
+        assert!(result.is_err());
+        assert_eq!(ChannelStatus::ChannelHandleInvalid, result.unwrap_err())
+    }
+
+    #[tokio::test]
+    async fn test_hosted_channel_read_no_message() {
+        let channel_handle = ChannelHandle::LookupData as i32;
+        let mut wasm_state = create_test_wasm_state();
+        // Assumes LookupData has a channel.
+        let result = wasm_state.channel_read(channel_handle, 0, 0);
+        assert!(result.is_err());
+        assert_eq!(ChannelStatus::ChannelEmpty, result.unwrap_err());
+    }
+
+    #[tokio::test]
+    async fn test_hosted_channel_read_channel_closed() {
+        let channel_handle = ChannelHandle::LookupData as i32;
+        let mut wasm_state = create_test_wasm_state();
+        // Assumes LookupData has a channel.
+        let extension_endpoints = &mut wasm_state.extensions_endpoints;
+        // Remove the endpoint of the runtime closes one endpoint of the channel.
+        extension_endpoints.remove(&LOOKUP_ABI_FUNCTION_NAME.to_string());
+        let result = wasm_state.channel_read(channel_handle, 0, 0);
+        assert!(result.is_err());
+        assert_eq!(
+            ChannelStatus::ChannelEndpointDisconnected,
+            result.unwrap_err()
+        );
+    }
+
+    #[tokio::test]
+    async fn test_hosted_channel_read_ok() {
+        let channel_handle = ChannelHandle::LookupData as i32;
+        let message = b"A Test Message: Value";
+        let mut wasm_state = create_test_wasm_state();
+
+        // Write message into Lookup endpoint.
+        let endpoint = wasm_state
+            .extensions_endpoints
+            .get(&LOOKUP_ABI_FUNCTION_NAME.to_string())
+            .unwrap();
+        let result = endpoint.sender.send(message.to_vec().clone()).await;
+        assert!(result.is_ok());
+
+        // Guess some memory addresses in linear Wasm memory.
+        let dest_ptr_ptr: AbiPointer = 100;
+        let dest_len_ptr: AbiPointer = 150;
+
+        let result = wasm_state.channel_read(channel_handle, dest_ptr_ptr, dest_len_ptr);
+        assert!(result.is_ok());
+
+        // Get dest_len from dest_len_ptr.
+        let dest_len: u32 =
+            LittleEndian::read_u32(&wasm_state.get_memory().get(dest_len_ptr, 4).unwrap());
```

Comment:
I thought we have some helper functions for this sort of read / conversion on WasmState? If not, is it worth adding them?

Reply:
I only found some for `write_u32/buffer_to_wasm_memory` but no read. 

Makes sense though, because I am not sure where we would need reading from Wasm memory apart from this test, so I would not add them at the moment.